### ISSUE TITLE
Is expand with filter supported?

### ISSUE THREAD

**Issue by Makison** [object/Simple.OData.Client#200]
Hi,

I have 1 to many relation product - productPrices
I need to get Products with at least 1 active price, and expand only active prices
below code get correct products, but return all ProductPrices including not active prices
```
 client.For<Product>().Top(10).Skip(0)
 .Filter(p => p.ProductPrices.Any(pp => pp.IsActive == true))
.Expand(pp => pp.ProductPrices)
.FindEntriesAsync();
```
query string:
`/odata/products?$filter=ProductPrices/any(x1:x1/IsActive eq true)&$skip=0&$top=10&$expand=ProductPrices`
response:
```
      "Id":1,"ManufacturerId":0,"UnitId":0,"BarCode":"0123450123","Code":"000000001","Name":"product 1","Description":"product 1 description","Created":"2016-02-07T01:47:43.14Z","IsActive":true,"ProductPrices":[
        {
          "Id":1,"ProductId":1,"Price":1001,"IsActive":false,"Created":"2016-02-07T01:47:43.253Z"
        },{
          "Id":2,"ProductId":1,"Price":11,"IsActive":true,"Created":"2016-02-07T01:47:43.253Z"
        }
      ]
```

what I'm trying to get (working as url):
`/odata/products?$filter=ProductPrices/any(x1:x1/IsActive eq true)&$skip=0&$top=10&$expand=ProductPrices($filter=IsActive eq true)`

response:
   ```
   "Id":1,"ManufacturerId":0,"UnitId":0,"BarCode":"0123450123","Code":"000000001","Name":"product 1","Description":"product 1 description","Created":"2016-02-07T01:47:43.14Z","IsActive":true,"ProductPrices":[
        {
          "Id":1,"ProductId":1,"Price":1001,"IsActive":true,"Created":"2016-02-07T01:47:43.253Z"
        }
      ]
```

as option add select:
`$expand=ProductPrices($filter=IsActive eq true;$select=Price)`

Thanks

**Comment by object** [comment]
Thank you for the issue report. That's an interesting one. The URL generated by the library is like this:

/odata/products?$filter=ProductPrices/any(x1:x1/IsActive eq true)&$skip=0&$top=10&$expand=ProductPrices

However, what you observed is that result contains both active and not active product prices. In other words, the filter is ignored.

Event though you managed to build another URL that retrieve correct results I wonder whether this the library's or service's fault: from my understanding the above URL is correct and should result in filtering product prices. I don't recall out of my head that there must be two filters attached to this type of query. I can investigate this further. What OData service are you using?

**Comment by Makison** [comment]
Hi,

I'm using ASP.NET Web API OData v4 Endpoint (System.Web.OData.ODataController). 

First I thought the same as you it should filter out ProductPrices but when I faced this issue, I realized that from logical point of view it works correctly - We specifying filter for Product and it returns Products containing at least one Active ProductPrice. It doesn't return products that don't have active price - so we get what we actually asked, and further more in other place I actually need this behaviour. So we should be able to filter expanded items separately
  
Second URL add filter fon the expanded PrductPrice. Example 92 in the doc
http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part2-url-conventions/odata-v4.0-errata02-os-part2-url-conventions-complete.html#_Toc406398162
 
Thanks

**Comment by object** [comment]
Thank you for the clarification. It seems to make sense. I will look into it.

**Comment by object** [comment]
I have investigated this issue and apparently it's harder to get it fixed than I first thought. I need to extend the supported syntax to recognize expressions like Expand(x =&gt; x.ProductPrices.Where(y =&gt; y.ProductPrices.IsActive)) and this will require significant changes to the expression parser.

I marked this issue as 5.0 but hope to get back to it soon.

**Comment by Makison** [comment]
Thank you!

**Comment by jmfronsee** [comment]
Probably no chance this is fixed yet?

**Comment by object** [comment]
No, as I wrote in earlier comment, this change requires major revision of some parts so it will come in version 5 (no date is set unfortunately due to my full engagement in other project).

**Comment by JoschaMetze** [comment]
I've implemented a prototype version to handle this situation: https://github.com/Hochfrequenz/Simple.OData.Client
If you're interested I can provide a pull request. The syntax is not as short as I'd like it to be but I haven't found a better version yet, maybe you can provide assistance with that.

Current example:
```c#
var person = await _client
                .For("People")
                .Key("russellwhyte")
                .Expand("Friends", "Trips/PlanItems")
                .Filter(new List<KeyValuePair<string, string>>() {
                    new KeyValuePair<string, string>("Friends", "LastName eq 'Ketchum'"),
                    new KeyValuePair<string, string>("Trips", "TripId eq 0"),
                    new KeyValuePair<string, string>("Trips/PlanItems", "PlanItemId eq 12") })
                .OrderBy("Friends/LastName")
                .FindEntryAsync();
            //Assert.Equal(3, person.Trips.Count());
            Assert.Single((person["Friends"] as IEnumerable<IDictionary<string, object>>));
            Assert.Single((person["Trips"] as IEnumerable<IDictionary<string, object>>));
            Assert.Single((person["Trips"] as IEnumerable<IDictionary<string, object>>).First()["PlanItems"] as IEnumerable<IDictionary<string, object>>);
```

This yields the ODATA URL: 
```
People('russellwhyte')?$expand=Friends($orderby=LastName;$filter=LastName eq 'Ketchum'),Trips($expand=PlanItems($filter=PlanItemId eq 12);$filter=TripId eq 0)
```
I've implemented a ODataExpression version as well (see test case for an example).

**Comment by object** [comment]
Thank you for the suggestion. It looks like it enables providing filters as key-value pairs. Since the library includes API based on LINQ expressions, it is very tempted to complement your implementation with LINQ-based syntax. I know parsing LINQ expression trees is not trivial, but without it changes feel incomplete. Let me think about whether it's worth to extend the API just by making this addition.

**Comment by GauchoCoder** [comment]
Any update on this issue? I am also looking to filter expanded properties, similar to above
`$expand=ProductPrices($filter=IsActive eq true)`.

**Comment by DeminLi** [comment]
have the $expand=ProductPrices($filter=IsActive eq true) been supported in version 5.10.0.1, yet? Anyone can share a working codes? I tried
`client.Select(d => new 
                {
                    ProductPrices = d.ProductPrices.Where(x => x.IsActive eq true)
                }`

**Comment by D1mitry** [comment]
Any update on this issue? Is it possible to have Expand method for client which will take string as a parameter? I mean like this
```
client.For<Product>().Top(10).Skip(0)
.Expand("ProductPrices($filter=IsActive eq true)")
.FindEntriesAsync();
```

**Comment by D1mitry** [comment]
Ok, I've got a working version, here is a sample:
```
client.For<Product>().Top(10).Skip(0)
.QueryOptions("$expand=ProductPrices($filter=IsActive eq true)")
.FindEntriesAsync();
```

**Comment by weitzhandler** [comment]
This repo is no longer maintained, check out ODataConnectedService

**Comment by maximpashuk** [comment]
@davidnmbond @object Any update on this?
Using .Filter inside .Expand will allow to recude network traffic between ODATA server and client.

**Comment by maximpashuk** [comment]
also I think issue #136 also can be closed since .select inside .expand is also supported in 5.26.0
